### PR TITLE
Add Tree-Sitter support

### PR DESCRIPTION
**Pull request by kjinho**
#63

### CODE REVIEW

File: org.gnu.emacs.json
Comment (muep) [inline org.gnu.emacs.json:109]: Not sure why this had to be included, at least to me it seems that our builds still work with the previous mechanism.

Sure we perhaps could still switch to this instead of a submodule, but I think it would be good to ship this separately, or just omit this for now if the build still works.

**Review by muep — CHANGES_REQUESTED** [review]
Thanks @kjinho for working on adding tree-sitter support. I feel it makes sense to have this, as long as we know what we are getting into.

Since Emacs has optional features depending on many kinds of things, many of them are not enabled at this time. Every time we a dependency is added and bundled, this means that the package maintainers need to be responsible for that as well.

I have not used tree-sitter myself, but on surface it sounds like a fairly impactful thing to enable. But could you summarize here a bit, what becomes possible if we ship the library and enable the configuration option? Also, do you expect any downsides to taking this action?

Also since I do not use this myself, I would feel much more confident applying this if I knew that you (or someone else interested) would intend to look after the dependency and submit updates when it needs updating.

And finally, it would be useful if some contributor uses that feature, because I would probably not know if the library is broken in some fashion in some point.

File: org.gnu.emacs.json
Comment (kjinho) [reply org.gnu.emacs.json:49]: It installs only the library, not the CLI tool. 

(I'll address your other questions shortly!)

File: org.gnu.emacs.json
Comment (kjinho) [reply org.gnu.emacs.json:109]: That's sensible. Let me try to build it with the submodule, and I'll report back.

Comment (kjinho) [reply org.gnu.emacs.json:109]: Okay, I've confirmed that building with the submodule (instead of a specified commit) currently works. There is a benefit to pinning to a specific, known-to-be-working state of the git repository, but I'm not going to insist that we use one approach over the other. As both work, I leave it to your judgment!

**Comment by kjinho** [comment]
> Thanks @kjinho for working on adding tree-sitter support. I feel it makes sense to have this, as long as we know what we are getting into.
> 

Thank you @muep for the thoughtful comments, and I appreciate your work on maintaining this package.

> Since Emacs has optional features depending on many kinds of things, many of them are not enabled at this time. Every time we a dependency is added and bundled, this means that the package maintainers need to be responsible for that as well.
> 
> I have not used tree-sitter myself, but on surface it sounds like a fairly impactful thing to enable. But could you summarize here a bit, what becomes possible if we ship the library and enable the configuration option? Also, do you expect any downsides to taking this action?
> 

These are excellent points, and I will admit that I put together this patch because I started using tree-sitter in Emacs some time ago but only recently switched to using Emacs via Flatpak. For my use case, there is not a substantial benefit to using the various tree-sitter modes in lieu of the more traditional major modes. But tree-sitter enables the implementation interesting features such as structural editing and navigation—[combobulate](https://github.com/mickeynp/combobulate) is one package that offers these features—, and I suspect that tree-sitter will replace regexps as the preferred method of creating major modes in the future.

The main downside that I see for adding support here is the increased cost of maintaining this added complexity to the package, even if it is modest. 

> Also since I do not use this myself, I would feel much more confident applying this if I knew that you (or someone else interested) would intend to look after the dependency and submit updates when it needs updating.
> 

I am happy to keep an eye on this and submit updates as necessary.

> And finally, it would be useful if some contributor uses that feature, because I would probably not know if the library is broken in some fashion in some point.

I'm pretty happy with using Emacs via Flatpak (big thanks to you and the other contributors!), and assuming that tree-sitter support is added, I will be using it.

Thanks again for reviewing the pull request, and let me know if you have any other questions.

**Comment by muep** [comment]
Hi @kjinho, sorry for me dropping the ball again, it has been slightly stressed in the past few weeks, so I failed to find space for working on this.

I think it would be good to pull this in if you are still interested, and the the latest diff seems otherwise ok in my view, but I wonder if the pgtk switch should be done separately.

Also since you seemed to be on top of also doing the 29.3 update, maybe you would be interested in being added as a maintainer for the package? I'd expect it might go smoother with more than one active maintainer?

**Comment by kjinho** [comment]
@muep No problem, whatsoever. I know how busy things can get, and I didn't want to have the pull request constantly pending while I tinkered with my own fork. 

I'll try to do separate pull requests for the various changes: (1) add tree-sitter support, (2) update to emacs 29.3, (3) switch to git for the wrapper, and (4) add pgtk (Wayland) support. You can then pick and choose what you want to merge.

I think you're doing a fine job, and I'm happy to continue contributing pull requests as appropriate!

**Comment by muep** [comment]
I did the 29.3 right away on top of the 29.2 which I had been testing because it was a small-ish change and supposedly urgent. Those other two would still be welcome.

**Pull request closed without merging**
